csharp/src/Ice/Communicator.cs

```diff
@@ -931,7 +924,7 @@ public Reference CreateReference(string s, string? propertyPrefix = null)
                                 throw new FormatException($"no argument provided for -p option `{s}'");
                             }
 
-                            protocol = Util.StringToProtocol(argument);
+                            protocol = ProtocolExtensions.Parse(argument);
```

Comment:
Parse is not an extension method (an extension method takes the type as `this` parameter). We could put this internal static method on Util rather than ProtocolExtensions ... a Protocol-related class sounds more logical to me, and there is only one, ProtocolExtensions.

Then an enum is better than a struct, because the enumerators are real constants. With a struct, the "enumerators" are static readonly variables, so you can't reference them from other static readonly variables (well, you can, but won't get the expected result). Making Encoding an enum in C# wouldn't be very elegant.

Reply:
Yes, I missed that the it was a simple static method.

What do you mean by you won't get the expected result?

Given that Protocol and Encoding are public APIs, as a user I would expect them to be similar... and given that we have a custom string representation, it could be convenient that ToString provides the preferred string representation.

Reply:
With static readonly variables referencing each other, you get 0.0 instead of the expected value. This is why the existing code (prior to this PR) uses the pattern:
```
public static readonly Protocol CurrentProtocol =
            new Protocol(Ice1Definitions.ProtocolMajor, Ice1Definitions.ProtocolMinor);

        public static readonly Encoding CurrentProtocolEncoding =
            new Encoding(Ice1Definitions.ProtocolEncodingMajor,
                                Ice1Definitions.ProtocolEncodingMinor);

        public static readonly Encoding CurrentEncoding =
            new Encoding(EncodingDefinitions.EncodingMajor, EncodingDefinitions.EncodingMinor);
```
This code is referencing real constants (the major/minor byte constants), not the static readonly variables, due to order-of-initialization issue. It's cleaner and safer with an enum.